pageserver/src/layered_repository.rs

```diff
@@ -2063,11 +2073,39 @@ impl LayeredTimeline {
             let (key, lsn, value) = x?;
 
             if let Some(prev_key) = prev_key {
-                if key != prev_key && writer.is_some() {
-                    let size = writer.as_mut().unwrap().size();
-                    if size > target_file_size {
-                        new_layers.push(writer.take().unwrap().finish(prev_key.next())?);
-                        writer = None;
+                if writer.is_some() {
+                    let size = writer.as_ref().unwrap().size();
+                    if lsn < writer.as_ref().unwrap().lsn_range.start
+                        || (prev_key != key && size > target_file_size)
+                        || size > target_file_size * 2
+                    {
+                        if prev_key == key {
+                            let final_lsn_range = Range {
+                                start: lsn_range.start,
+                                end: lsn,
+                            };
+                            new_layers.push(
+                                writer
+                                    .take()
+                                    .unwrap()
+                                    .finish(final_lsn_range, prev_key.next())?,
+                            );
```

Comment:
This doesn't work. You cannot set `final_lsn_range.lsn.end` to 'lsn`. You might have already written WAL records in to the delta file with LSN > `lsn`. So you end up with a layer that contains WAL records outside the LSN range that's in the file's header.

Reply:
I see...
So looks like there are four possible ways:
1. Avoid directory serialization. Looks like in this case we wil not have problem with to larger number of duplicates of the same key.
2. Two passes (at first pass we calculate slit boundaries).
3. Truncate file (and B-Tree) at the beginning of last key range and creating layer with just one key.
4.  Modify B-Tree to to support posting lists. In this case we can easily calculate  occupied by each key and so make decision to split layer before start processing values of the key.

Any other proposal?

Reply:
I see.
So the simplest solution is not to address splitting keys at compaction phase at all. So just half of this patch which handles overflow of inmem layer is needed. As far as number of L1 layers is limited (right now 6), then size of layer file after compaction can not be larger than 6_checkpoint_distance = 6*256MB = 1.5Mb. But for some reasons it is now more than 2Gb.

Reply:
I create yet another PR #1927 implementing the following approach:
- Persist image layers when any of the following two conditions is true:
  1. LSN range >= `checkpoint_distance`
  2. Layer size > `checkpoint_distance`
- Perform L0 level compaction when any  of the following two conditions is true
  1. Number of L0 layers >= `compaction_threshold`
  2. Any of L0 layers contains LSN range < `checkpoint_distance`

In other words, if layer contains large BLOBs (like serialized directories) which are not present in WAL and so been persisted according to the second (size) criteria, then we force recompaction to prevent creation of huge ( `compaction_threshold * checkpoint_distance` ) delta layers.

I also added your modified version of test_hot_update test (test_dup_key.py)  which should reproduce problem with large layers without large schema. But I failed to reproduce the problem with this test.

Reply:
I found out that this criteria (LSN range < `checkpoint_distance`) is not correct (it is true for most layers).
So I have to add `get_max_key_range()` method to Layer (implemented only for DeltaLayer) and force compaction of L0 layers if sum of  `get_max_key_range()`  exceeds target_file_size.